include/fmt/format.h

```diff
@@ -1830,28 +1838,58 @@ auto write_float(OutputIt out, const DecimalFP& fp,
       if (num_zeros <= 0 && fspecs.format != float_format::fixed) num_zeros = 1;
       if (num_zeros > 0) size += to_unsigned(num_zeros) + 1;
     }
-    auto grouping = digit_grouping<Char>(loc, fspecs.locale);
-    size += to_unsigned(grouping.count_separators(significand_size));
-    return write_padded<align::right>(out, specs, size, [&](iterator it) {
-      if (sign) *it++ = static_cast<Char>(data::signs[sign]);
-      it = write_significand<Char>(it, significand, significand_size,
-                                   fp.exponent, grouping);
-      if (!fspecs.showpoint) return it;
-      *it++ = decimal_point;
-      return num_zeros > 0 ? detail::fill_n(it, num_zeros, zero) : it;
-    });
+    // this is lambda, because GCC 11 cannot handle digit_grouping usages at
+    // compile-time
+    auto runtime_path = [&]() {
+      auto grouping = digit_grouping<Char>(loc, fspecs.locale);
```

Comment:
I suggest moving constexpr handling to `digit_grouping`. What exactly is the problem with it?

Reply:
If I understand this correctly, GCC fails when `digit_grouping` is used in any of `if` branch directly, doesn't matter if it's under `is_constant_evaluated()` check and in `else` clause. Why does it fail - probably because `std::string` inside `digit_grouping` is not constexpr. Why does it fail in a constant-evaluated context with `digit_grouping` not reachable - IDK, from my experience, a simple `is_constant_evaluated()` check was enough to avoid this kind of problem. Even MSVC handles code correctly with `digit_grouping` usage under a `is_constant_evaluated()` check.

Reply:
Let's parameterize `digit_grouping` on `is_constant_evaluated` and provide a fallback that does no grouping if it's true. This will keep uses of `digit_grouping` simple and once `std::string` becomes fully constexpr we'll be able to drop parametrization.

Reply:
If you mean that `digit_grouping` should have a template parameter, it probably won't work because of the nature of `is_constant_evaluated` - it always returns `true` when invoked at compile-time. That's why we cannot write
`if constexpr (is_constant_evaluated())`.

Reply:
You are right, parametrization on `is_constant_evaluated` won't work but we could still have some dispatch mechanism that chooses whether to use the fallback or real `digit_grouping`. Something along the lines of https://godbolt.org/z/5rbcEEK87.

Reply:
Don't underestimate GCC with its ability to find usages of `std::string` in this particular case - https://godbolt.org/z/7fTdhnsPd. Let's also thank GCC for this glorious error message. 😉

Reply:
Interesting but I think something like this should work: https://godbolt.org/z/9he3Yobo4

Reply:
Done